buildkite/bazel-central-registry/bcr_presubmit.py

```diff
@@ -257,6 +262,89 @@ def run_test(repo_location, task_config_file, task):
+    )
+
+    # Check if any of the source.json, MODULE.bazel, or patch files are changed for an existing module.
+    NO_CHANGE_FILE_PATTERNS = [
+        r"modules\/([^\/]+)\/([^\/]+)\/source.json",
+        r"modules\/([^\/]+)\/([^\/]+)\/MODULE.bazel",
+        r"modules\/([^\/]+)\/([^\/]+)\/patches",
+    ]
+    changed_modules = []
+    for line in output.decode("utf-8").split():
+        for p in NO_CHANGE_FILE_PATTERNS:
+            s = re.match(p, line)
+            if s:
+                bazelci.eprint(line, "was changed")
+                changed_modules.append(s.groups())
+    if changed_modules:
+        error("Existing modules should not be changed:\n" + "\n".join([f"{name}@{version}" for name,version in changed_modules]))
+    else:
+        bazelci.eprint("No existing module was changed.")
+
+
+def validate_files_outside_of_modules_dir_are_not_modififed(modules):
+    # If no modules are changed at the same time, then we don't need to perform this check.
+    if not modules:
+        return
+    bazelci.print_expanded_group("Checking if any file changes outside of modules/")
+    output = subprocess.check_output(
+        ["git", "diff", "main...HEAD", "--name-only", "--pretty=format:", ":!modules"]
+    ).decode("utf-8").strip()
+    if output:
+        error("The following files should not be changed when adding a new module version:\n" + output)
+    else:
+        bazelci.eprint("Nothing changed outside of modules/")
+
+
+def should_bcr_validation_block_presubmit(modules):
+    if not modules:
+        return
+    bazelci.print_expanded_group("Running BCR validations:")
+    returncode = subprocess.run(
+        ["python3", "./tools/bcr_validation.py"] + [f"--check={name}@{version}" for name, version in modules]
+    ).returncode
+    # When a BCR maintainer view is required, the script should return 42.
+    if returncode == 42:
+        return True
+    if returncode != 0:
+        raise BcrPipelineException("BCR validation failed!")
+    return False
+
+
+def should_wait_bcr_maintainer_review(modules):
+    """Validate the changes and decide whether the presubmit should wait for a BCR maintainer review.
+    Returns False if all changes look good and the presubmit can proceed.
+    Returns True if the changes should block follow up presubmit jobs until a BCR maintainer triggers them.
+    Throws an error if the changes violate BCR policies or the BCR validations fail.
+    """
+    # If existing modules are changed, fail the presubmit.
+    validate_existing_modules_are_not_modified()
+
+    # If files outside of the modules/ directory are changed, fail the presubmit.
+    validate_files_outside_of_modules_dir_are_not_modififed(modules)
+
+    # Run BCR validations on target modules and decide if the presubmit jobs should be blocked.
+    if should_bcr_validation_block_presubmit(modules):
+        return True
+
+    return False
+
+
+def upload_jobs_to_pipeline(pipeline_steps):
```

Comment:
To allow other output to be printed in stdout, I'll have to upload the jobs by directly calling the buildkite-agent instead of by print it out and pipe stdin to it.